File: backend/sms-api/src/main/java/com/sms/api/service/ActivityService.java
package com.sms.api.service;

import com.sms.api.dto.activity.ActivityDto;
import com.sms.api.dto.activity.CreateActivityRequest;
import com.sms.api.entity.Activity;
import com.sms.api.repository.ActivityRepository;
import com.sms.core.exception.ResourceNotFoundException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.UUID;

@Service
@Transactional
public class ActivityService {

    private final ActivityRepository activityRepository;

    public ActivityService(ActivityRepository activityRepository) {
        this.activityRepository = activityRepository;
    }

    @Transactional(readOnly = true)
    public List<ActivityDto> list(UUID schoolId) {
        return activityRepository.findBySchoolIdOrderByNameAsc(schoolId)
            .stream().map(this::toDto).toList();
    }

    public ActivityDto create(UUID schoolId, CreateActivityRequest req) {
        Activity a = new Activity();
        a.setSchoolId(schoolId);
        a.setName(req.name());
        a.setCategory(req.category());
        a.setCoach(req.coach());
        a.setSchedule(req.schedule());
        a.setCapacity(req.capacity() > 0 ? req.capacity() : 30);
        a.setStatus(req.status() != null ? req.status() : "ACTIVE");
        return toDto(activityRepository.save(a));
    }

    public ActivityDto update(UUID schoolId, UUID id, CreateActivityRequest req) {
        Activity a = activityRepository.findByIdAndSchoolId(id, schoolId)
            .orElseThrow(() -> new ResourceNotFoundException("Activity", id));
        a.setName(req.name());
        a.setCategory(req.category());
        a.setCoach(req.coach());
        a.setSchedule(req.schedule());
        if (req.capacity() > 0) a.setCapacity(req.capacity());
        if (req.status() != null) a.setStatus(req.status());
        return toDto(activityRepository.save(a));
    }

    public void delete(UUID schoolId, UUID id) {
        Activity a = activityRepository.findByIdAndSchoolId(id, schoolId)
            .orElseThrow(() -> new ResourceNotFoundException("Activity", id));
        activityRepository.delete(a);
    }

    private ActivityDto toDto(Activity a) {
        return new ActivityDto(a.getId(), a.getName(), a.getCategory(),
            a.getCoach(), a.getSchedule(), a.getCapacity(), a.getStatus());
    }
}

File: frontend/src/app/core/services/activity.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';

export interface ActivityDto { id: string; name: string; category: string; coach: string | null; schedule: string | null; capacity: number; status: string; }
export interface CreateActivityRequest { name: string; category: string; coach?: string; schedule?: string; capacity: number; status?: string; }

@Injectable({ providedIn: 'root' })
export class ActivityService {
  private http = inject(HttpClient);
  private base = `${environment.apiUrl}/activities`;
  list(): Observable<ActivityDto[]>                                    { return this.http.get<ActivityDto[]>(this.base); }
  create(req: CreateActivityRequest): Observable<ActivityDto>          { return this.http.post<ActivityDto>(this.base, req); }
  update(id: string, req: CreateActivityRequest): Observable<ActivityDto> { return this.http.put<ActivityDto>(`${this.base}/${id}`, req); }
  delete(id: string): Observable<void>                                 { return this.http.delete<void>(`${this.base}/${id}`); }
}
